### PR TITLE
[SPARK-33618][CORE] Use hadoop-client instead of hadoop-client-api to make hadoop-aws work

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This reverts commit SPARK-33212 (cb3fa6c9368e64184a5f7b19688181d11de9511c) mostly with three exceptions:
1. `SparkSubmitUtils` was updated recently by SPARK-33580
2. `resource-managers/yarn/pom.xml` was updated recently by SPARK-33104 to add `hadoop-yarn-server-resourcemanager` test dependency.
3. Adjust `com.fasterxml.jackson.module:jackson-module-jaxb-annotations` dependency in K8s module which is updated recently by SPARK-33471.

### Why are the changes needed?

According to [HADOOP-16080](https://issues.apache.org/jira/browse/HADOOP-16080) since Apache Hadoop 3.1.1, `hadoop-aws` doesn't work with `hadoop-client-api`. It fails at write operation like the following.

**1. Spark distribution with `-Phadoop-cloud`**

```scala
$ bin/spark-shell --conf spark.hadoop.fs.s3a.access.key=$AWS_ACCESS_KEY_ID --conf spark.hadoop.fs.s3a.secret.key=$AWS_SECRET_ACCESS_KEY
20/11/30 23:01:24 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Spark context available as 'sc' (master = local[*], app id = local-1606806088715).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.1.0-SNAPSHOT
      /_/

Using Scala version 2.12.10 (OpenJDK 64-Bit Server VM, Java 1.8.0_272)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.read.parquet("s3a://dongjoon/users.parquet").show
20/11/30 23:01:34 WARN MetricsConfig: Cannot locate configuration: tried hadoop-metrics2-s3a-file-system.properties,hadoop-metrics2.properties
+------+--------------+----------------+
|  name|favorite_color|favorite_numbers|
+------+--------------+----------------+
|Alyssa|          null|  [3, 9, 15, 20]|
|   Ben|           red|              []|
+------+--------------+----------------+


scala> Seq(1).toDF.write.parquet("s3a://dongjoon/out.parquet")
20/11/30 23:02:14 ERROR Executor: Exception in task 0.0 in stage 2.0 (TID 2)/ 1]
java.lang.NoSuchMethodError: org.apache.hadoop.util.SemaphoredDelegatingExecutor.<init>(Lcom/google/common/util/concurrent/ListeningExecutorService;IZ)V
```

**2. Spark distribution without `-Phadoop-cloud`**
```scala
$ bin/spark-shell --conf spark.hadoop.fs.s3a.access.key=$AWS_ACCESS_KEY_ID --conf spark.hadoop.fs.s3a.secret.key=$AWS_SECRET_ACCESS_KEY -c spark.eventLog.enabled=true -c spark.eventLog.dir=s3a://dongjoon/spark-events/ --packages org.apache.hadoop:hadoop-aws:3.2.0,org.apache.hadoop:hadoop-common:3.2.0
...
java.lang.NoSuchMethodError: org.apache.hadoop.util.SemaphoredDelegatingExecutor.<init>(Lcom/google/common/util/concurrent/ListeningExecutorService;IZ)V
  at org.apache.hadoop.fs.s3a.S3AFileSystem.create(S3AFileSystem.java:772)
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CI.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
cc @sunchao

**Comment by dongjoon-hyun** [comment]
This is for testing the feasibility as one of the option.

File: pom.xml [hunk 1]
```diff
@@ -1405,24 +1371,8 @@
             <groupId>com.sun.jersey.contribs</groupId>
             <artifactId>*</artifactId>
           </exclusion>
-          <exclusion>
-            <groupId>org.apache.hadoop</groupId>
-            <artifactId>hadoop-yarn-server-resourcemanager</artifactId>
-          </exclusion>
         </exclusions>
       </dependency>
-      <!--
```
Comment (HyukjinKwon) [inline pom.xml:1414]: Ahh .. this is going to be complicated I guess. In the fork in the company I work for, we still keep this change (SPARK-33104) alone. We did not port back SPARK-33212 yet. This is from SBT upgrade actually. 

It broke the test with Hadoop 2 so we still needed this change (SPARK-33104). Let's revert it together first as you proposed for now. I will take a look again to fix it in.

**Comment by HyukjinKwon** [comment]
I am good with reverting this first. I will take a look separately for SPARK-33104. Presumably the tests will fail with Hadoop 2.

**Comment by HyukjinKwon** [comment]
@dongjoon-hyun do you mind fixing the PR title and description to contain SPARK-33104 and https://github.com/apache/spark/commit/10bd42cd475eea8d5e5689e770e4773cebf62374?

**Comment by sunchao** [comment]
Yes I'm fine for reverting this first while we searching for other solutions. Let's hope we can still ship this in Spark 3.1 release.

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon and @sunchao .
This is still testing to check the feasibility to revert~ This PR will wait until next Monday. :)

BTW, I'll update the PR title and description.

File: pom.xml (hunk 1 above)
Comment (HyukjinKwon) [reply pom.xml]: @dongjoon-hyun feel free to revert this too if that's easier for you. I don't mind taking a look again. I will leave it to you.

**Comment by dongjoon-hyun** [comment]
Hi, All.
To investigate this more during Apache Spark 3.1 QA timeframe, I filed a new JIRA.
We have a few approaches including this and https://github.com/apache/spark/pull/30556 .

**Comment by dongjoon-hyun** [comment]
Hi, @HyukjinKwon .
Could you review this PR?

**Comment by dongjoon-hyun** [comment]
Also, cc @viirya , @dbtsai , @sunchao , @srowen , @AngersZhuuuu , @mridulm , @tgravescs .

**Review by viirya — APPROVED** [review]
Looks okay as I compared with SPARK-33212 (cb3fa6c).

**Review by HyukjinKwon — APPROVED** [review]
I already reviewed this actually. Was wondering which one you guys prefer. LGTM

**Comment by HyukjinKwon** [comment]
Merged to master.

File: dev/deps/spark-deps-hadoop-3.2-hive-2.3 [hunk 2]
```diff
@@ -50,13 +55,30 @@ datanucleus-api-jdo/4.2.4//datanucleus-api-jdo-4.2.4.jar
...
 gson/2.2.4//gson-2.2.4.jar
 guava/14.0.1//guava-14.0.1.jar
-hadoop-client-api/3.2.0//hadoop-client-api-3.2.0.jar
-hadoop-client-runtime/3.2.0//hadoop-client-runtime-3.2.0.jar
+guice-servlet/4.0//guice-servlet-4.0.jar
+guice/4.0//guice-4.0.jar
+hadoop-annotations/3.2.0//hadoop-annotations-3.2.0.jar
+hadoop-auth/3.2.0//hadoop-auth-3.2.0.jar
+hadoop-client/3.2.0//hadoop-client-3.2.0.jar
+hadoop-common/3.2.0//hadoop-common-3.2.0.jar
```
Comment (srowen) [inline dev/deps/spark-deps-hadoop-3.2-hive-2.3:71]: This pulls a ton more code into Spark now, like the whole client... hm, is this going to affect the hadoop-provided distro? it also downgrades some versions above which may be harmless. We really need this just for hadoop-aws?

Comment (HyukjinKwon) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:71]: Oh, @srowen this is basically a revert. There was an issue found of shading hadoop client it was reverted as a safe choice. A fix is in progress.

Comment (srowen) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:71]: Ah OK, nevermind. I am not following closely.

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya , @HyukjinKwon and @srowen !

File: dev/deps/spark-deps-hadoop-3.2-hive-2.3 [hunk 3]
```diff
@@ -116,14 +142,30 @@ jline/2.14.6//jline-2.14.6.jar
...
 jta/1.1//jta-1.1.jar
 jul-to-slf4j/1.7.30//jul-to-slf4j-1.7.30.jar
+kerb-admin/1.0.1//kerb-admin-1.0.1.jar
+kerb-client/1.0.1//kerb-client-1.0.1.jar
+kerb-common/1.0.1//kerb-common-1.0.1.jar
+kerb-core/1.0.1//kerb-core-1.0.1.jar
+kerb-crypto/1.0.1//kerb-crypto-1.0.1.jar
+kerb-identity/1.0.1//kerb-identity-1.0.1.jar
+kerb-server/1.0.1//kerb-server-1.0.1.jar
+kerb-simplekdc/1.0.1//kerb-simplekdc-1.0.1.jar
```
Comment (yaooqinn) [inline dev/deps/spark-deps-hadoop-3.2-hive-2.3:162]: just for curiosity, does spark has a chance to play the role of KDC at runtime?

Comment (HyukjinKwon) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:162]: This is actually a revert. It was added in https://github.com/apache/spark/commit/ce7ba2e98e0a3b038e881c271b5905058c43155b#diff-e45e1eee8dcfd7eaf8a013cec02b67806da3edeabe0f195ac6b4402f67d4b6dcR146

Comment (yaooqinn) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:162]: looks like the original PR does not handle any transitive artifact exclusion at all 😸

**Pull request closed without merging**